### ISSUE TITLE
Duplicated API on initialization

### ISSUE THREAD

**Issue by eliecer2323** [aws-controllers-k8s/community#894]
When I followed the official AWS guide to use API Gateway with AWS EKS using the ACK Controller I get like bug in the step when I need to create the API.

Here is the yaml configuration that I use and its result in AWS API Gateway

```
cat > apigw-api.yaml<<EOF
apiVersion: apigatewayv2.services.k8s.aws/v1alpha1
kind: API
metadata:
  name: apitest-private-nlb
spec:
  body: '{
              "openapi": "3.0.1",
              "info": {
                "title": "ack-apigwv2-import-test-private-nlb",
                "version": "v1"
              },
              "paths": {
              "/\$default": {
                "x-amazon-apigateway-any-method" : {
                "isDefaultRoute" : true,
                "x-amazon-apigateway-integration" : {
                "payloadFormatVersion" : "1.0",
                "connectionId" : "$(kubectl get vpclinks.apigatewayv2.services.k8s.aws \
  nlb-internal \
  -o jsonpath="{.status.vpcLinkID}")",
                "type" : "http_proxy",
                "httpMethod" : "GET",
                "uri" : "$(aws elbv2 describe-listeners \
  --load-balancer-arn $(aws elbv2 describe-load-balancers \
  --region $AGW_AWS_REGION \
  --query "LoadBalancers[?contains(DNSName, '$(kubectl get service authorservice \
  -o jsonpath="{.status.loadBalancer.ingress[].hostname}")')].LoadBalancerArn" \
  --output text) \
  --region $AGW_AWS_REGION \
  --query "Listeners[0].ListenerArn" \
  --output text)",
               "connectionType" : "VPC_LINK"
                  }
                }
              },
              "/meta": {
                  "get": {
                    "x-amazon-apigateway-integration": {
                       "uri" : "$(aws elbv2 describe-listeners \
  --load-balancer-arn $(aws elbv2 describe-load-balancers \
  --region $AGW_AWS_REGION \
  --query "LoadBalancers[?contains(DNSName, '$(kubectl get service echoserver \
  -o jsonpath="{.status.loadBalancer.ingress[].hostname}")')].LoadBalancerArn" \
  --output text) \
  --region $AGW_AWS_REGION \
  --query "Listeners[0].ListenerArn" \
  --output text)",
                      "httpMethod": "GET",
                      "connectionId": "$(kubectl get vpclinks.apigatewayv2.services.k8s.aws \
  nlb-internal \
  -o jsonpath="{.status.vpcLinkID}")",
                      "type": "HTTP_PROXY",
                      "connectionType": "VPC_LINK",
                      "payloadFormatVersion": "1.0"
                    }
                  }
                }
              },
              "components": {}
        }'
EOF
kubectl apply -f apigw-api.yaml
```

![Screen Shot 2021-08-09 at 7 37 07 PM](https://user-images.githubusercontent.com/26780188/128791134-2dbd066d-fd7c-4980-bfef-14c66ed4d99a.png)

If you run the `kubectl get apis` command you'll notice that it is only connected to one of this two, then you can remove the other one and continue working with the connected one, but I think there is a problem on the first creation of the API because it is making it twice

**Comment by vijtrip2** [comment]
Hi @eliecer2323 , Thanks for bringing this to our attention. I will try to reproduce and hopefully resolve next week.

**Comment by aandrushchenko** [comment]
same here, excerpt from contoller logs:
```
2021-08-16T09:21:53.751Z	INFO	ackrt	created new resource	{"kind": "API", "namespace": "ack-system", "name": "*masked*", "generation": 1, "is_adopted": false, "arn": null}
...
...
2021-08-16T09:21:54.240Z	INFO	ackrt	created new resource	{"kind": "API", "namespace": "ack-system", "name": "*masked*", "generation": 1, "is_adopted": false, "arn": null}
```

**Comment by vijtrip2** [comment]
The latest release v0.3.0 for apigatewayv2 controller fixes this bug. Please use the latest helm chart to resolve this issue. Thank you for the patience.

Release: https://github.com/aws-controllers-k8s/apigatewayv2-controller/releases/tag/v0.0.3
New Controller Image: public.ecr.aws/aws-controllers-k8s/apigatewayv2-controller:v0.0.3 
New Helm Chart: public.ecr.aws/aws-controllers-k8s/apigatewayv2-chart:v0.0.3

**Comment by eliecer2323** [comment]
when I do `helm chart pull public.ecr.aws/aws-controllers-k8s/apigatewayv2-chart:v0.0.3` it gives me this error

```
v0.0.3: Pulling from public.ecr.aws/aws-controllers-k8s/apigatewayv2-chart
Error: unexpected status code [manifests v0.0.3]: 403 Forbidden
```

**Comment by eliecer2323** [comment]
It was hard to me for updating from the v0.0.2 to the new v0.0.3, I was trying to uninstall everything from the previous and installing it again from the new one but for some reason it keep loading the deployment and never finish, I assume it is because of the permissions but I didn't find what should I change in the permissions, I used this guide previously (https://aws.amazon.com/blogs/containers/integrate-amazon-api-gateway-with-amazon-eks/) for installing the v0.0.2 and it uses this permissions, could you tell me what should I change in it? or if there is some other updated guide? Thank you.

```
curl -O https://raw.githubusercontent.com/aws-samples/amazon-apigateway-ingress-controller-blog/Mainline/apigw-ingress-controller-blog/ack-iam-policy.json

aws iam create-policy \
  --policy-name ACKIAMPolicy \
  --policy-document file://ack-iam-policy.json

eksctl create iamserviceaccount \
  --attach-policy-arn=arn:aws:iam::${AGW_ACCOUNT_ID}:policy/ACKIAMPolicy \
  --cluster=$AGW_EKS_CLUSTER_NAME \
  --namespace=kube-system \
  --name=ack-apigatewayv2-controller \
  --override-existing-serviceaccounts \
  --region $AGW_AWS_REGION \
  --approve
```

**Comment by vijtrip2** [comment]
@eliecer2323 , there may have been some changes in the parameters required by controllers from v0.0.2 to v0.0.3
I am going to go over this blog, execute it myself for v0.0.3 and create a set of instructions which needs to be updated in that blog.
I will share my findings with you sometime next week.

However here are some more instructions for you on how the installation for ACK controllers work. While I go through the blog post, you can use following to try and debug things yourself too.
* https://aws-controllers-k8s.github.io/community/user-docs/install/
* Getting Started on https://github.com/aws-controllers-k8s/sagemaker-controller/blob/main/README.md

**Comment by eliecer2323** [comment]
I appreciate your support @vijtrip2 , I am checking the links you shared me but the second link is for AWS SageMaker (does it have something with APIGateway2?), and the first link I think it hasn't been updated too, in the installation step after downloaded the repo it seems to be using the old name of the repo, also it doesn't seem to contain any configuration about policies for installing and get running it
![Screen Shot 2021-08-22 at 12 09 49 PM](https://user-images.githubusercontent.com/26780188/130363897-1e5d2b5b-e0b2-4062-a32e-bb7698733de4.png)

I think the command should look like these, because that's the new repository name.
```
export ACK_K8S_NAMESPACE=ack-system

kubectl create namespace $ACK_K8S_NAMESPACE

helm install --namespace $ACK_K8S_NAMESPACE $SERVICE-chart \
    $CHART_EXPORT_PATH/$SERVICE-chart
```

Anyway it never gets ready when I search into
<img width="1233" alt="Screen Shot 2021-08-22 at 3 41 55 PM" src="https://user-images.githubusercontent.com/26780188/130369440-d92f7919-57a4-4bda-905f-be0abf134699.png">

and with the old version it works
<img width="1209" alt="Screen Shot 2021-08-22 at 3 47 33 PM" src="https://user-images.githubusercontent.com/26780188/130369614-5fb2c4f7-fbe8-479e-a457-9f79a3136422.png">

**Comment by vijtrip2** [comment]
> "the second link is for AWS SageMaker (does it have something with APIGateway2?)"

I should have mentioned it. It is for sagemaker controller but it lays down the steps on how to install ACK service controller. So I thought that could help you find anything you were doing differently.

> "the first link I think it hasn't been updated too, in the installation step after downloaded the repo it seems to be using the old name of the repo"

That "ack-$SERVICE-controller" in that helm install command is a user-input. It can be any string. so that should not be an issue.

-------------------------

About the IAM permissions, you should be able to use the same permissions mentioned here in your IRSA role. https://raw.githubusercontent.com/aws-samples/amazon-apigateway-ingress-controller-blog/Mainline/apigw-ingress-controller-blog/ack-iam-policy.json

I am going to follow the blog post for setting up load balancer etc and then use installation instructions in ACK docs for installing APIGWv2 controller v0.3.0 . I will keep this issue updated with my findings.

**Comment by vijtrip2** [comment]
I did try to follow the blog, mixed with installation instructions from sagemaker guide. Running into a very tiny issue with IRSA but when that's fixed, I will post the instructions here and then work on filling the gaps in ACK installation instructions.

**Comment by vijtrip2** [comment]
I was able to successfully install v0.3.0 controller(I will paste instructions here) but the original issue of duplicate API is still present. It was not present earlier when I tested the v0.3.0 controller with all the upgrades.

The root-cause is that the API takes some time to get created on AWS side. When k8s controller retries, finds no existing API and it attempts to create another one. This only happens on rare cases when CreateAPI latencies are high.

I am diving deep to find the correct solution for it.

**Comment by vijtrip2** [comment]
I had some issues in trying to create LoadBalancer service in my EKS cluster so could not follow the whole blog post but I did successfully install apigatewayv2 controller v0.3.0 and created a test api(how i found the duplicate bug is still there.) Mostly these instructions are from sagemaker guide and worked for installing apigatewayv2 controller.

I will focus on duplicate bug now. Find the instructions below

```
export AWS_DEFAULT_REGION=us-east-1
export AWS_ACCOUNT_ID=$(aws sts get-caller-identity --query 'Account' --output text)
export CLUSTER_NAME=sample

// Did not run this. I already have a cluster 'sample'
eksctl create cluster \
  --name $CLUSTER_NAME \
  --region $AWS_DEFAULT_REGION \
  --managed


export VPC_ID=$(aws eks describe-cluster \
  --name $CLUSTER_NAME \
  --region $AWS_DEFAULT_REGION  \
  --query "cluster.resourcesVpcConfig.vpcId" \
  --output text)

helm repo add eks https://aws.github.io/eks-charts && helm repo update

> "eks" has been added to your repositories
> Update Complete. ⎈Happy Helming!⎈

kubectl apply -k "github.com/aws/eks-charts/stable/aws-load-balancer-controller//crds?ref=master"
> customresourcedefinition.apiextensions.k8s.io/ingressclassparams.elbv2.k8s.aws created
> customresourcedefinition.apiextensions.k8s.io/targetgroupbindings.elbv2.k8s.aws created

helm install aws-load-balancer-controller \
  eks/aws-load-balancer-controller \
  --namespace kube-system \
  --set clusterName=$CLUSTER_NAME \
  --set serviceAccount.create=false \
  --set serviceAccount.name=aws-load-balancer-controller \
  --set vpcId=$VPC_ID\
  --set region=$AWS_DEFAULT_REGION
> NAME: aws-load-balancer-controller
> LAST DEPLOYED: Sun Aug 22 20:44:03 2021
> NAMESPACE: kube-system
> STATUS: deployed
> REVISION: 1
> TEST SUITE: None
> NOTES:
> AWS Load Balancer controller installed!


kubectl get nodes
> NAME                             STATUS   ROLES    AGE   VERSION
> ip-192-168-56-233.ec2.internal   Ready    <none>   44d   v1.19.6-eks-49a6c0

=== APIGATEWAYV2 controller setup start ============

export OIDC_PROVIDER_URL=$(aws eks describe-cluster --name $CLUSTER_NAME --region $AWS_DEFAULT_REGION \
--query "cluster.identity.oidc.issuer" --output text | cut -c9-)

echo $OIDC_PROVIDER_URL
> oidc.eks.us-east-1.amazonaws.com/id/D46C603EA3D759EA476D017EB4*****


printf '{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "Federated": "arn:aws:iam::'$AWS_ACCOUNT_ID':oidc-provider/'$OIDC_PROVIDER_URL'"
      },
      "Action": "sts:AssumeRoleWithWebIdentity",
      "Condition": {
        "StringLike": {
          "'$OIDC_PROVIDER_URL':aud": "sts.amazonaws.com",
          "'$OIDC_PROVIDER_URL':sub": [
            "system:serviceaccount:ack-system:ack-*"
          ]
        }
      }
    }
  ]
}
' > ./trust.json

export OIDC_ROLE_NAME=ack-controller-role-$CLUSTER_NAME

aws --region $AWS_DEFAULT_REGION iam create-role --role-name $OIDC_ROLE_NAME --assume-role-policy-document file://trust.json

curl -O https://raw.githubusercontent.com/aws-samples/amazon-apigateway-ingress-controller-blog/Mainline/apigw-ingress-controller-blog/ack-iam-policy.json

aws iam create-policy \
  --policy-name ACKIAMPolicy \
  --policy-document file://ack-iam-policy.json

aws --region $AWS_DEFAULT_REGION iam attach-role-policy --role-name $OIDC_ROLE_NAME --policy-arn arn:aws:iam::$AWS_ACCOUNT_ID:policy/ACKIAMPolicy

export IAM_ROLE_ARN_FOR_IRSA=$(aws --region $AWS_DEFAULT_REGION iam get-role --role-name $OIDC_ROLE_NAME --output text --query 'Role.Arn')

export HELM_EXPERIMENTAL_OCI=1
export SERVICE=apigatewayv2
export RELEASE_VERSION=v0.0.3
export CHART_EXPORT_PATH=/tmp/chart
export CHART_REPO=public.ecr.aws/aws-controllers-k8s/$SERVICE-chart
export CHART_REF=$CHART_REPO:$RELEASE_VERSION
mkdir -p $CHART_EXPORT_PATH
__pw=$(aws ecr-public get-login-password --region us-east-1)
echo "$__pw" | helm registry login -u AWS --password-stdin public.ecr.aws
> Login succeeded

helm chart pull $CHART_REF
> v0.0.3: Pulling from public.ecr.aws/aws-controllers-k8s/apigatewayv2-chart

helm chart list
> REF                                                         	NAME                       	VERSION            	DIGEST 	SIZE    	CREATED
> public.ecr.aws/aws-controllers-k8s/apigatewayv2-chart:v0.0.3	apigatewayv2-chart         	v0.0.3             	95efebd	10.5 KiB	19 seconds

helm chart export $CHART_REF --destination $CHART_EXPORT_PATH
cd $CHART_EXPORT_PATH/$SERVICE-chart
yq e '.aws.region = env(AWS_DEFAULT_REGION)' -i values.yaml
yq e '.aws.account_id = env(AWS_ACCOUNT_ID)' -i values.yaml
yq e '.serviceAccount.annotations."eks.amazonaws.com/role-arn" = env(IAM_ROLE_ARN_FOR_IRSA)' -i values.yaml
cd -

kubectl apply -f $CHART_EXPORT_PATH/$SERVICE-chart/crds

export ACK_K8S_NAMESPACE=ack-system
helm install -n $ACK_K8S_NAMESPACE --create-namespace --skip-crds ack-$SERVICE-controller \
 $CHART_EXPORT_PATH/$SERVICE-chart

kubectl get pods -n $ACK_K8S_NAMESPACE
> NAME                                                              READY   STATUS    RESTARTS   AGE
> ack-apigatewayv2-controller-apigatewayv2-chart-67b47d7847-6zkjq   1/1     Running   0          2m33s

kubectl create -f api.yaml
api.apigatewayv2.services.k8s.aws/my-api created

kubectl get apis
NAME     AGE
my-api   8s
```

**Comment by vijtrip2** [comment]
I think i found the issue why your controller was not starting for version v0.0.3. The account_id was added as a required parameter in newer version of ACK controller and the path to install controllers was also changed.

Replace the commands in section "The ACK team has published Helm chart to install ACK on Amazon ECR Public Gallery. Install ACK using Helm:" with following commands and it should work then.

```
export HELM_EXPERIMENTAL_OCI=1
export SERVICE=apigatewayv2
export RELEASE_VERSION=v0.0.3
export CHART_EXPORT_PATH=/tmp/chart
export CHART_REPO=public.ecr.aws/aws-controllers-k8s/apigatewayv2-chart
export CHART_REF=$CHART_REPO:$RELEASE_VERSION
mkdir -p $CHART_EXPORT_PATH

__pw=$(aws ecr-public get-login-password --region us-east-1)
echo "$__pw" | helm registry login -u AWS --password-stdin public.ecr.aws
> Login succeeded

helm chart pull $CHART_REF
> v0.0.3: Pulling from public.ecr.aws/aws-controllers-k8s/apigatewayv2-chart
> ref:     public.ecr.aws/aws-controllers-k8s/apigatewayv2-chart:v0.0.3
> digest:  26a2cedf48960080a4690bcf68fc887945731deb03d8fd36ce438fa5e0cb1f2b
> size:    10.5 KiB
> name:    apigatewayv2-chart
> version: v0.0.3
> Status: Chart is up to date for public.ecr.aws/aws-controllers-k8s/apigatewayv2-chart:v0.0.3

helm chart export $CHART_REF --destination $CHART_EXPORT_PATH
> ref:     public.ecr.aws/aws-controllers-k8s/apigatewayv2-chart:v0.0.3
> digest:  4fba9d5362da03131b2026abaf2bc04a2d9ae3014b619b9bd87a1f358b2348db
> size:    10.5 KiB
> name:    apigatewayv2-chart
> version: v0.0.3
> Exported chart to /tmp/chart/apigatewayv2-chart/

helm install ack-${SERVICE}-controller \
  $CHART_EXPORT_PATH/${SERVICE}-chart \
  --namespace kube-system \
  --set serviceAccount.create=false \
  --set aws.region=$AGW_AWS_REGION
  --set aws.account_id=$AGW_ACCOUNT_ID
```

**Comment by edijsdrezovs** [comment]
Same issue for me. Tested with latest version `v0.0.3`

**Comment by vijtrip2** [comment]
I have found the root cause of the problem. Working on finding a good solution now.

```
I have found that API Gatewayv2 controller, when it creates an API, patches spec as per create output and triggers another reconciliation with metadata.Generation=2.
During this immediate second reconciliation, sometimes ReadOne returns NotFound because status.APIID is still not updated into the etcd. Status is updated at the end of the first reconciliation loop.
This NotFound triggers another creation and k8s user end up with duplicate APIs in some edge cases.

The issue is very hard to reproduce as it occurs rarely but causes resource leak.
```

**Comment by ops-hummus** [comment]
It happens to me also, on high latency environment. It created 3 resources and not 1 :\

**Comment by vijtrip2** [comment]
I did the investigation on this and found no bug in the controller code. 

The issue is arising because the identifier for `APIGatewayv2` is in `Status` part of k8s resource. K8s treats `Status` as the observed state of the resource and for API Gateway finding observed state requires the identifier which is present in the status. Hence if the `Status` is not present, `APIGatewayv2 controller` has no knowledge whether a resource was created or not.

In my investigation, I created 50 APIs and found that 4-5 were duplicated. I looked at logs for those apis (I added additional logs) and found that the desired resource being read right after `kc.Status.Patch()` call has empty status so API Gateway controller has no choice but to create another API on next reconciliation and add it's ID to k8s resource `Status`.

Interesting thing is that `kc.Status().Patch()` returned no error in above case to indicate that status patching had failed. All the calls were successful.

---------------

What's next?
* I will dive deeper into whether Get calls for k8s client are cached and return stale copy of resource in some cases?
* I have create a PR for v0.0.4 release of `APIGatewayv2` controller. It does not solve this issue but contains other improvement in ACK runtime(v0.0.13)

---------------

In my personal opinion, I would love for this issue to go away completely but because of nature of API Gateway controller (where identifier is in the k8s resource status), it needs some more investigation.
However, if those duplicate APIs are being created, they should not cause any harm to your workload. Once the reconciler stabilizes with one `API Id` it will not create more APIs.
And about the duplicate APIs, they are just dangling APIs. You will not be charged for them unless any invocation happens for them. Cleaning them up will be an issue as you will have to go to console or use aws-cli, but hopefully soon I will find a solution for this problem completely.

Thank you for your patience on this!

**Comment by vijtrip2** [comment]
APIGatewayv2 v0.0.4 release is out.
Controller Image: public.ecr.aws/aws-controllers-k8s/apigatewayv2-controller:v0.0.4
Helm Chart: public.ecr.aws/aws-controllers-k8s/apigatewayv2-chart:v0.0.4

**Comment by vijtrip2** [comment]
An update on this issue:

>In my investigation, I created 50 APIs and found that 4-5 were duplicated. I looked at logs for those apis (I added additional logs) and found that the desired resource being read right after kc.Status.Patch() call has empty status so API Gateway controller has no choice but to create another API on next reconciliation and add it's ID to k8s resource Status.

> Interesting thing is that kc.Status().Patch() returned no error in above case to indicate that status patching had failed. All the calls were successful.

With more testing I have found out that root cause is the cached result returned from `k8sClient.Get()` call, which returns a cached desired resource(contains no status) and it triggers the duplicate API creation.

Having found the actual root cause now, I am working on a fix and this bug will be fixed in release v0.0.5(coming soon)
Appreciate your patience!

**Comment by vijtrip2** [comment]
APIGatewayv2 v0.0.5 release is out. This fixes the duplicate API creation bug.
Controller Image: public.ecr.aws/aws-controllers-k8s/apigatewayv2-controller:v0.0.5
Helm Chart: public.ecr.aws/aws-controllers-k8s/apigatewayv2-chart:v0.0.5

--------

closing the issue but please reopen if you think the bug is not fixed. 
Cheers!
